numpy/ma/core.py

```diff
@@ -7400,22 +7401,195 @@ def mask_rowcols(a, axis=None):
       fill_value=1)
 
     """
-    a = array(a, subok=False)
+    a = asarray(a)
+    # rather than generalizing this weird API to ND, have the user call the
+    # simpler API
     if a.ndim != 2:
-        raise NotImplementedError("mask_rowcols works for 2D arrays only.")
+        raise NotImplementedError(
+            "mask_rowcols works for 2D arrays only, use mask_extend_axis "
+            "instead (which takes a different axis argument)."
+        )
+    # compatibility logic to accept the weird input that was allowed before
+    axes = ()
+    if not axis:
+        axes += (1,)
+    if axis in [None, 1, -1]:
+        axes += (0,)
+    return mask_extend_axis(a, axes)
+
+
+def mask_extend_axis(a, axis):
```

Comment:
Suggestions for a better name appreciated

Reply:
I think I had similar, probably not very good ideas, on the mailing list. But what about: `extend_mask(ed)_along_axis`. Sorry for randomly coming back to this, but my personal opinion is we just need to has out the correct name.

The way multiple axis work may be a bit specific to matrix multiply though, so maybe we should just make it private for now to move things ahead. With a comment that we may make it public in the future?